core/shared/src/main/scala/scalaz/zio/ZManaged.scala

```diff
@@ -73,93 +213,516 @@ final case class ZManaged[-R, +E, +A](reserve: ZIO[R, E, Reservation[R, E, A]])
+  /**
+   * Returns an effect that dies with the specified `Throwable`.
+   * This method can be used for terminating a fiber because a defect has been
+   * detected in the code.
+   */
+  final def die(t: Throwable): ZManaged[Any, Nothing, Nothing] = halt(Cause.die(t))
+
+  /**
+   * Returns an effect that dies with a [[java.lang.RuntimeException]] having the
+   * specified text message. This method can be used for terminating a fiber
+   * because a defect has been detected in the code.
+   */
+  final def dieMessage(message: String): ZManaged[Any, Throwable, Nothing] = die(new RuntimeException(message))
+
+  /**
+   * Returns an effect from a [[scalaz.zio.Exit]] value.
+   */
+  final def done[E, A](r: Exit[E, A]): ZManaged[Any, E, A] =
+    ZManaged.fromEffect(ZIO.done(r))
+
+  /**
+   * Accesses the whole environment of the effect.
+   */
+  final def environment[R]: ZManaged[R, Nothing, R] =
+    ZManaged.fromEffect(ZIO.environment)
+
+  /**
+   * Returns an effect that models failure with the specified error.
+   * The moral equivalent of `throw` for pure code.
+   */
+  final def fail[E](error: E): ZManaged[Any, E, Nothing] =
+    halt(Cause.fail(error))
+
+  /**
+   * Returns an effect that performs the outer effect first, followed by the
+   * inner effect, yielding the value of the inner effect.
+   *
+   * This method can be used to "flatten" nested effects.
+    **/
+  final def flatten[R, E, A](zManaged: ZManaged[R, E, ZManaged[R, E, A]]): ZManaged[R, E, A] =
+    zManaged.flatMap(scala.Predef.identity)
+
+  /**
+   * Applies the function `f` to each element of the `Iterable[A]` and
+   * returns the results in a new `List[B]`.
+   *
+   * For a parallel version of this method, see `foreachPar`.
+   */
+  final def foreach[R, E, A1, A2](as: Iterable[A1])(f: A1 => ZManaged[R, E, A2]): ZManaged[R, E, List[A2]] =
+    as.foldRight[ZManaged[R, E, List[A2]]](succeed(Nil)) { (a, m) =>
+      f(a).zipWith(m)(_ :: _)
+    }
+
+  /**
+   * Applies the function `f` to each element of the `Iterable[A]` in parallel,
+   * and returns the results in a new `List[B]`.
+   *
+   * For a sequential version of this method, see `foreach`.
+   */
+  final def forEachPar[R, E, A1, A2](
+    as: Iterable[A1]
+  )(
+    f: A1 => ZManaged[R, E, A2]
+  ): ZManaged[R, E, List[A2]] =
+    as.foldRight[ZManaged[R, E, List[A2]]](ZManaged.succeed(List())) {
+      case (a, man) =>
+        f(a).zipWithPar(man)(_ :: _)
+    }
+
+  /**
+   * Applies the function `f` to each element of the `Iterable[A]` in parallel,
+   * and returns the results in a new `List[B]`.
+   *
+   * Unlike `foreachPar`, this method will use at most up to `n` fibers.
+   */
+  final def forEachParN[R, E, A1, A2](
+    n: Long
+  )(
+    as: Iterable[A1]
+  )(
+    f: A1 => ZManaged[R, E, A2]
+  ): ZManaged[R, E, List[A2]] =
+    for {
+      semaphore <- ZManaged.fromEffect(Semaphore.make(n))
+      bs <- forEachPar(as) { a =>
+             ZManaged {
+               f(a).reserve.map {
```

Comment:
Not sure here if the `withPermit` should be on the `reserve` or on the `acquire` 🧐

Reply:
Good point. We might also do something like this:
```
  final def foreachParN_[R, E, A](n: Long)(as: Iterable[A])(f: A => ZManaged[R, E, _]): ZManaged[R, E, Unit] =
    for {
      semaphore <- ZManaged.fromEffect(Semaphore.make(n))
      _ <- foreachPar_(as) { a =>
            ZManaged {
              ZIO.uninterruptibleMask { restore =>
                semaphore.acquire *> restore(f(a).reserve).foldCauseM(
                  cause => semaphore.release *> ZIO.halt(cause),
                  { case Reservation(acquire, release) =>
                    ZIO.succeed(Reservation(restore(acquire).ensuring(semaphore.release), release))
                  }
                )
              }
            }
          }
    } yield ()
```

Reply:
See my reply at foldM

Reply:
I opened a ticket for the design discussion: https://github.com/scalaz/scalaz-zio/issues/846
I also played around with the implementations a bit and I think I have one that should be satisfactory:
```
  /**
   * Applies the function `f` to each element of the `Iterable[A]` and runs
   * produced effects in parallel, discarding the results.
   *
   * Unlike `foreachPar_`, this method will use at most up to `n` fibers.
   */
  final def foreachParN_[R, E, A](
    n: Long
  )(
    as: Iterable[A]
  )(
    f: A => ZManaged[R, E, _]
  ): ZManaged[R, E, Unit] =
  ZManaged[R, E, Unit] {
    Ref.make[ZIO[R, Nothing, Any]](IO.unit).map { finalizers =>
      Reservation(
        for {
          queue  <- Queue.unbounded[(A, Promise[E, Unit])]
          worker  = queue.take.flatMap { case (a, prom) =>
            ZIO.uninterruptibleMask { restore =>
              f(a).reserve
                .flatMap(res => finalizers.update(fs => res.release *> fs).const(res))
                .flatMap(res => restore(res.acquire))
            }.foldCauseM(
              _.failureOrCause.fold(prom.fail, prom.halt),
              _ => prom.succeed(())
            )
          }
          _      <- ZIO.foreach(1L to n)(_ => worker.forever.fork)
          proms  <- ZIO.foreach(as) { a =>
            for {
              prom <- Promise.make[E, Unit]
              _    <- queue.offer((a, prom))
            } yield prom
          }
          _      <- ZIO.foreach_(proms)(_.await).ensuring(queue.shutdown)
        } yield (),
        ZIO.flatten(finalizers.get)
      )
    }
  }
```

While working on this I noticed a general issue with the *par methods: Implementing them in terms of zipWithPar as done in ZIO might not be the smartest idea. As raceWith uses interruption under the hood in the case of ZManaged all reservations and therefore all release actions will always run, even when the first effect immediately fails.
This issue does not exist with a queue based implementation as above, but I'm not sure what approach is better for the regular parN methods

Reply:
Is this really a problem? In an ideal implementation, we would start all reservations at the exact same time; the fact that we may be able to avoid starting some of them is only a sign we're starting them up in linear fashion, since once reservation begins, it must complete or fail.

Further performance work will try to get much closer to "at the same time" than, e.g. `foreachPar` right now.